### PR TITLE
Reuse connections (variation of PR #51)

### PR DESCRIPTION
**Pull request by exvito**
Following @Lukasa's comment, here is the PR.

### CODE REVIEW

File: requests_ntlm/requests_ntlm.py [hunk 1]
```diff
@@ -36,33 +35,25 @@ def __init__(self, username, password, session=None):
...
 
-        adapter = self.adapter
-        if self.session:
-            session = self.session()
-            if session:
-                adapter = session.get_adapter(response.request.url)
+        # Consume content and release the original connection
+        # to allow our new request to reuse the same one.
+        response.content
+        response.raw.release_conn()
```
Comment (Lukasa) [inline requests_ntlm/requests_ntlm.py]: I *think* this logic is insufficient. It works fine in the single-threaded case (which might be good enough), but urllib3's connection-pooling logic will mean that releasing the connection in this way could lead to a failure to get it back.

@sigmavirus24, is it worth trying to explicitly hold on to the connection here?

**Comment by exvito** [comment]
I'm not 100% into the low level details, but I can confirm that as long as I use a pool manager adequate for my use case (see #56), the multi-threaded behaviour is ok - while failing consistently without that pool manager.

(Humbly underlining "my use case") :)

**Comment by exvito** [comment]
@Lukasa,

Here is an update (because I never feel confortable not understanding things 100%, and went digging deeper):
* This PR does improve connection reuse during the authentication process.
* The `response.raw.release_conn()` above seems "by-the-book" in the sense that it returns the kept-alive connection to the pool, which (apparently) must be done because there is a very loose coupling between HTTPAdapter (used to send subsequent requests) and the underlying PoolManager.
* This PR does not, however, ensure two fully independent app-level requests (two threads, for example) are NTLM-authenticated correctly: not unless the pools managed by the PoolManager use block=True and hold a maximum of one conncection. Again, this stems from the loose coupling between HTTPAdapater and the PoolManager: on send, the HTTPAdapater does `conn = self.get_connection(..)` which, in turn, ends up calling PoolManager's `connection_from_url(..)`.

More:
* I raised the parallelism in my use case and quickly found failures with this PR, with PR #51 with v0.1.0;  with either the default PoolManager or the variation I shared in #56.
* For some reason, using a PoolManager with block=1 and a forcing its underlying pools to have at most one connection, lead me to some hangs which I haven't diagnosed yet (but then again, by code base is somewhat complex).

Summary:
* I'd like to understand this even better (in particular, why I'm getting hangs -- maybe I'll write a simpler test client that reproduces the failures I'm observing and go from there).
* If anyone can chime in with an idea or point out where I'm mistaken, that'd be great.
* Nonetheless in the single-threaded case (and in lots of other threaded cases, but not all) this PR does indeed improve connection reuse.

Again,
Thanks in advance for any input.

**Comment by Lukasa** [comment]
I think I'm generally positive on this change, though like you @exvito I think I want to try to get a feel for the multithreaded use case. Regardless, if @sigmavirus24 and @mastahyeti are up for it I'm happy to merge this.

**Comment by exvito** [comment]
Quick update:
* Went digging further.
* Given my understanding for the need of connection-level authenticatin, suspected requests' bundled HTTPDigestAuth would be a good candidate to suffer from the same limitaions.
* Created a multi-threaded test client that shows this to be the case.
* Ended up finding requests' issue 2334 (https://github.com/kennethreitz/requests/issues/2334) with a sample test client: same failure, same underlying issue.

Extra:
* Found out requests bundled urllib3 (and urllib3 by itself) includes a contributed NTLMConnectionPool. This takes a different approach at addressing NTLM authentication.

Next:
* Regarding this PR: waiting on your feedback, of course.
* Regarding my need for multi-threaded NTLM authentication:
  * May give the NTLMConnectionPool a test run even though it does not integrate as nicely with my codebase, since it does not work at the requests/auth API level.
  * If my understanding is correct, the current requests authentication model, HTTPAdapter and the underlying urllib3's PoolManager design provide no way to ensure that a particular requests authenticator gets the same connection that got an initial 401 (unless pools block and have at most one connection, I believe). Fixing this while keeping backwards compatibility seems feasible, but is something non-immediate and, of course, outside of the scope of this PR and project. :)

**Comment by Lukasa** [comment]
Thanks for the detailed analysis @exvito. I think you've got a good overview of this here. Just waiting on @sigmavirus24 and @mastahyeti, but I'm :+1: on merging this as-is.

**Comment by mastahyeti** [comment]
> I think you've got a good overview of this here. Just waiting on @sigmavirus24 and @mastahyeti, but I'm :+1: on merging this as-is.

I'm pretty far out of the loop on requests/requests-ntlm stuff these days. If you and @sigmavirus24 and :+1: I'd be happy to merge this and cut a release.

**Comment by sigmavirus24** [comment]
I don't have the time at the moment to very closely read this, but I think there are a lot of fundamental flaws in this pull request.

For example, operations on the original request shouldn't be done via `response.request` if we have to do anything like that, it should not be a constant access via the response object because attribute access hurts in Python more than people realize. Further, I think we support a set of requests versions that means we can *not* assume that a `PreparedRequest` has a `copy` method. Admittedly, there isn't automated CI for this project, but that doesn't mean we can just break backwards compatibility randomly. I'm also confused as to why we're releasing the socket but then using `response.connection` (which if memory serves me right is the `urllib3.HTTPConnection` (which should have said socket). We're releasing it and then using it. I don't think that's what we actually want to do.

**Comment by Lukasa** [comment]
@sigmavirus24 Best to hold off until you have time to read the whole discussion: at least part of your feedback is heavily discussed in this thread. =)

**Comment by exvito** [comment]
@sigmavirus24,

Thanks for your input. I'll gladly standby until you have a chance to look at this more closely and further comment.

Quick notes as simple "brain-dump" so as to keep a reference of things for whenever you have that chance:
* Interested in "attribute access hurts in Python more than people realize" (performance? threads? something else?)
* Understand and agree with not breaking backwards compatibility.
* The "releasing the socket and then using `response.connection`" was inspired by the Digest authenticator which, at the time, I believed also authenticated connections -- I've since learned that that is not a fact.
* This code does indeed improve conncetion reuse, however.
* Last but not least, unfortunatelly, `response.connection` is an `requests.HTTPAdapter,` not a `urllib3.HTTPConnection`.

**Comment by Lukasa** [comment]
While there are concerns, I think for the moment it's more important that we resolve the known problems with requests-ntlm that this fixes.

The concern about what version of requests we support, and whether it supports `PreparedRequest.copy`, is a valid one, but I think bumping the minor version number and increasing the required version of requests is probably pretty safe. At this point, depending on at least requests v2.0 (which is when `PreparedRequest.copy` went in) feels safe.

**Pull request merged**